### PR TITLE
[server, protocol] GuardedCostCenter and AttributionId.parse/render (3/5)

### PR DESCRIPTION
**Pull request by geropl**
## Related Issue(s)

Context: #10325 

## How to test
 - `(cd components/server && yarn test)`

## Release Notes

```release-note
NONE
```

## Werft options:

- [ ] /werft with-preview

### CODE REVIEW

File: components/server/ee/src/workspace/gitpod-server-impl.ts [hunk 1]
```diff
@@ -2058,9 +2059,49 @@ export class GitpodServerEEImpl extends GitpodServerImpl {
...
+                    throw new ResponseError(ErrorCodes.NOT_FOUND, "Team not found");
+                }
+                const members = await this.teamDB.findMembersByTeam(team.id);
+                owner = { kind: "team", team, members };
+                break;
+            case "user":
+                owner = { kind: "user", userId };
+                break;
+            default:
+                throw new ResponseError(ErrorCodes.BAD_REQUEST, "Invalid attributionId");
```
Comment (easyCZ) [inline components/server/ee/src/workspace/gitpod-server-impl.ts:2100]:
```suggestion
                throw new ResponseError(ErrorCodes.BAD_REQUEST, "Invalid attributionId type");
```

File: components/server/ee/src/workspace/gitpod-server-impl.ts [hunk 2]
```diff
@@ -2058,9 +2059,49 @@ export class GitpodServerEEImpl extends GitpodServerImpl {
...
+        operation: ResourceAccessOp,
+    ): Promise<void> {
+        traceAPIParams(ctx, { userId, attributionId });
+
+        // TODO(gpl) We need a CostCenter entity (with a strong connection to Team or User) to properly to authorize access to these reports
+        // const costCenter = await this.costCenterDB.findByAttributionId(attributionId);
+        const parsedId = AttributionId.parse(attributionId);
+        if (parsedId === undefined) {
+            log.warn({ userId }, "Unable to parse attributionId", { attributionId });
+            throw new ResponseError(ErrorCodes.BAD_REQUEST, "Unable to parse attributionId");
```
Comment (easyCZ) [inline components/server/ee/src/workspace/gitpod-server-impl.ts:2083]: Nice status codes!

File: components/server/ee/src/workspace/gitpod-server-impl.ts [hunk 3]
```diff
@@ -2058,9 +2059,49 @@ export class GitpodServerEEImpl extends GitpodServerImpl {
...
+                owner = { kind: "team", team, members };
+                break;
+            case "user":
+                owner = { kind: "user", userId };
+                break;
+            default:
+                throw new ResponseError(ErrorCodes.BAD_REQUEST, "Invalid attributionId");
+        }
+
+        await this.guardAccess({ kind: "costCenter", /*subject: costCenter,*/ owner }, operation);
```
Comment (easyCZ) [inline components/server/ee/src/workspace/gitpod-server-impl.ts:2103]: Commented out section?

**Review by easyCZ — APPROVED** [review]
LGTM. Adding hold for the questions.

/hold

File: components/gitpod-protocol/src/attribution.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+    export function parse(s: string): UserAttributionId | TeamAttributionId | undefined {
+        if (!s) {
+            return undefined;
+        }
+        const parts = s.split(":");
+        if (parts.length !== 2) {
+            return undefined;
+        }
+        switch (parts[0]) {
```
[the remark this answers is not in the document]
Comment (geropl) [reply components/gitpod-protocol/src/attribution.ts:30]: It's covered with the default branch.

File: components/server/src/auth/resource-access.ts [hunk 5]
```diff
@@ -104,6 +106,24 @@ export interface GuardedTeam {
     members: TeamMemberInfo[];
 }
 
+export interface GuardedCostCenter {
+    kind: "costCenter";
+    //subject: CostCenter;
```
[the remark this answers is not in the document]
Comment (geropl) [reply components/server/src/auth/resource-access.ts:111]: It's meant to be a (strong) reminder to add it later (e.g., next week). Felt this was going out-of-scope for now.

**Pull request merged**